### PR TITLE
feat: add service discovery check

### PR DESCRIPTION
**Pull request by rarguelloF**
### What does this PR do?

This PR adds a new `service_discovery` corecheck (enabled by default) that scans processes looking for potential services the user can monitor.

### Motivation

New Service Discovery feature.

### CODE REVIEW

File: comp/forwarder/eventplatform/eventplatformimpl/epforwarder.go
Comment (jonbodner) [inline comp/forwarder/eventplatform/eventplatformimpl/epforwarder.go:216]: can you double-check with @yshapiro-57 or someone else on the telemetry team that this is the correct way to use their endpoint?

File: pkg/logs/client/http/destination.go
Comment (jonbodner) [inline pkg/logs/client/http/destination.go:222]: Did you intend to leave this here?

File: pkg/process/portlist/port.go
Comment (jonbodner) [inline pkg/process/portlist/port.go:1]: should this package be under servicediscovery?

File: pkg/collector/corechecks/servicediscovery/check.go
Comment (jonbodner) [inline pkg/collector/corechecks/servicediscovery/check.go:117]: use a `map[string]bool` for a set

File: pkg/collector/corechecks/servicediscovery/check.go
Comment (jonbodner) [inline pkg/collector/corechecks/servicediscovery/check.go:30]: use a map[string]bool for a set.

File: pkg/collector/corechecks/servicediscovery/check.go
Comment (jonbodner) [inline pkg/collector/corechecks/servicediscovery/check.go:46]: should this be a slice of strings since it could be > 1 process?

File: pkg/collector/corechecks/servicediscovery/discover_services_linux.go
Comment (jonbodner) [inline pkg/collector/corechecks/servicediscovery/discover_services_linux.go:1]: There are no tests. We need tests.

File: pkg/collector/corechecks/servicediscovery/discover_services_linux.go
Comment (jonbodner) [inline pkg/collector/corechecks/servicediscovery/discover_services_linux.go:133]: We should gather telemetry on these cases to find out how often it failed. I'd like to send back the command line, but the problem is going to be making sure that we don't capture any sensitive information.

File: pkg/collector/corechecks/servicediscovery/discover_services_linux.go
Comment (jonbodner) [inline pkg/collector/corechecks/servicediscovery/discover_services_linux.go:125]: `map[int]bool`

File: pkg/collector/corechecks/servicediscovery/discover_services_linux.go
Comment (jonbodner) [inline pkg/collector/corechecks/servicediscovery/discover_services_linux.go:163]: This is the reason why you should use map[T]bool for a set. The test to see if something matches becomes just `if aliveProcesses[pid]`

File: pkg/collector/corechecks/servicediscovery/discover_services_linux.go
Comment (jonbodner) [inline pkg/collector/corechecks/servicediscovery/discover_services_linux.go:87]: I think we had discussed getting rid of the ignoreMap because a process that opens up a listen port after the scan will end up being ignore.

File: pkg/collector/corechecks/servicediscovery/discover_services_linux.go
Comment (jonbodner) [inline pkg/collector/corechecks/servicediscovery/discover_services_linux.go:237]: I think we discussed a different algorithm that should be simpler.

- get map of ports that are listened on to associated PIDs
   - for each PID
      - if we haven’t seen it before, put on “potential” list
      - if it is on the “potential” list, add to “service” list, remove from “potential” list, send start-service event
      - if it is on the “service” list and it’s been 15 minutes, send heartbeat-service event
  - for each PID on the “service” list that wasn’t in the map of ports->PIDs
     - send end-service event
     - remove from “service” list
- sleep for 1 minute

File: pkg/collector/corechecks/servicediscovery/discover_services_linux.go
Comment (jonbodner) [inline pkg/collector/corechecks/servicediscovery/discover_services_linux.go:109]: should the poller be state in the linuxImpl struct?

File: pkg/collector/corechecks/servicediscovery/discover_services_linux.go
Comment (jonbodner) [inline pkg/collector/corechecks/servicediscovery/discover_services_linux.go:104]: I don't think this is implementing the right algorithm. We had talked about a simplification where we had a list of all processes and a list of (port,pid) pairs (I think I called it a map).

File: pkg/collector/corechecks/servicediscovery/discover_services_linux.go
Comment (jonbodner) [inline pkg/collector/corechecks/servicediscovery/discover_services_linux.go:28]: 1. why do these maps need to be concurrency-safe?
2. if we need them, we should use a generic thread-safe type, since the logic and key are identical; all that's different is the data stored in the map.

File: comp/forwarder/eventplatform/eventplatformimpl/epforwarder.go
Comment (rarguelloF) [reply comp/forwarder/eventplatform/eventplatformimpl/epforwarder.go:216]: 👍  adding @yshapiro-57 as reviewer.

File: pkg/logs/client/http/destination.go
Comment (rarguelloF) [reply pkg/logs/client/http/destination.go]: removed!

File: pkg/process/portlist/port.go
Comment (rarguelloF) [reply pkg/process/portlist/port.go:1]: moved under servicediscovery

File: pkg/collector/corechecks/servicediscovery/discover_services_linux.go
Comment (rarguelloF) [reply pkg/collector/corechecks/servicediscovery/discover_services_linux.go]: will take a look on how to do this

File: pkg/collector/corechecks/servicediscovery/discover_services_linux.go
Comment (rarguelloF) [reply pkg/collector/corechecks/servicediscovery/discover_services_linux.go]: fixed!

File: pkg/collector/corechecks/servicediscovery/discover_services_linux.go
Comment (rarguelloF) [reply pkg/collector/corechecks/servicediscovery/discover_services_linux.go]: I was previously spawning goroutines, but I'm not doing this anymore so I removed this code.

File: releasenotes/notes/add-service-discovery-corecheck-19f43e79510ae93b.yaml
Comment (buraizu) [inline releasenotes/notes/add-service-discovery-corecheck-19f43e79510ae93b.yaml:11]:
```suggestion
    Adds the new service_discovery check, which is currently only supported on Linux.
```

**Review by buraizu — APPROVED** [review]
Approving with a minor suggestion to the release note

File: pkg/collector/corechecks/servicediscovery/events.go
Comment (yshapiro-57) [inline pkg/collector/corechecks/servicediscovery/events.go:29]: It should be OK to not include the API version and request type in the payload since they are already included in the event. The spec in https://docs.google.com/document/d/1dc5KRIJjtONFjkWaITZF4i53UHgAjuextT6jyWmzJ-k/edit#heading=h.6ytopxdl5w22 went through a few iterations and when we added these fields to the outer struct, we might have forgotten to remove them from the inner one.

**Review by yshapiro-57 — COMMENTED** [review]
For the requests to be accepted by our event platform endpoint, they will need to include a header with an API key identifying the sender's org. I don't seem to see where we're injecting the header in the HTTP request, am I missing it?

You should also consider running an end-to-end test similar to the description of https://github.com/DataDog/dd-go/pull/131165 where you build and run the agent on a VM, and confirm that messages from the agent show up in the logs of our intake service in staging.

The JSON structure of the new event types looks reasonable to me other than the inline comment. Thank you for working on this.

File: pkg/collector/corechecks/servicediscovery/impl_linux.go
Comment (jonbodner) [inline pkg/collector/corechecks/servicediscovery/impl_linux.go:143]: in theory, it's possible that the pid won't be in the list, but the odds are astronomically small. We should use `p, ok := proces[pid]` and check the value of `ok` just to be safe.

File: pkg/collector/corechecks/servicediscovery/impl_linux.go
Comment (jonbodner) [inline pkg/collector/corechecks/servicediscovery/impl_linux.go:146]: Maybe we should put on the ignore list if it fails so we don't flood the log with process info failures.

File: pkg/collector/corechecks/servicediscovery/impl_linux.go
Comment (jonbodner) [inline pkg/collector/corechecks/servicediscovery/impl_linux.go:124]: since we got the service info when we detected the potential service, there shouldn't be a need to get it again.

File: pkg/collector/corechecks/servicediscovery/impl_linux.go
Comment (jonbodner) [inline pkg/collector/corechecks/servicediscovery/impl_linux.go:116]: why is this a pointer? same question for following line.

File: pkg/collector/corechecks/servicediscovery/impl_linux.go
Comment (jonbodner) [inline pkg/collector/corechecks/servicediscovery/impl_linux.go:88]: this is very confusing code. At first glance, a variable is created, a field in it is assigned a value, and then the variable is dead code. It works because events is a pointer to a value that was already assigned in e. 

The code is clearer if `getOrCreate` is rolled into the `addStarted` and `addStopped` functions:

```go
events, ok := e[svc.meta.Name]
if !ok {
    events = processEvents{}
}
events.started = append(events.started,svc)
e[svc.meta.Name] = events
```

I find this much easier to follow and the reduction in pointers makes less work for the garbage collector.

File: pkg/collector/corechecks/servicediscovery/impl_linux.go
Comment (jonbodner) [inline pkg/collector/corechecks/servicediscovery/impl_linux.go:190]: we should still send the start service events and ignore the end service events. That way we don't time out the row in REDAPL and the start time is kept up to date.

File: pkg/collector/corechecks/servicediscovery/events.go
Comment (jonbodner) [inline pkg/collector/corechecks/servicediscovery/events.go:39]: This field should be of type string, with three legal values:

- none
- packaged
- injected

I need to extract code from the injection library to provide these values.

File: pkg/collector/corechecks/servicediscovery/impl_linux_test.go
Comment (jonbodner) [inline pkg/collector/corechecks/servicediscovery/impl_linux_test.go:48]: For some reason, I can't get these tests to run on my Mac. Are there any special instructions with the mock library? I am not a fan of using them.

**Review by jonbodner — CHANGES_REQUESTED** [review]
I think things look mostly ok, but there are small issues to fix. Finer-grained tests on the functions and methods would be useful; see what I did for the language testing.

File: pkg/collector/corechecks/servicediscovery/impl_linux.go
Comment (rarguelloF) [reply pkg/collector/corechecks/servicediscovery/impl_linux.go:116]: changed to values 👍

File: pkg/collector/corechecks/servicediscovery/impl_linux_test.go
Comment (rarguelloF) [reply pkg/collector/corechecks/servicediscovery/impl_linux_test.go:48]: This test file has build tags `//go:build linux` since its using types / functions from `impl_linux.go` which has the same build tags (mockgen should not be the problem since its just autogenerating code).

 I'm running them in a debian VM (a docker container should also work) with the following command:

```
go test -v ./pkg/collector/corechecks/servicediscovery/ -tags=test,linux
```

File: pkg/collector/corechecks/servicediscovery/events.go
Comment (rarguelloF) [reply pkg/collector/corechecks/servicediscovery/events.go:29]: I still see these fields in the document (at least in the example). I can remove these fields, but just double-checking.

**Review by carlosroman — CHANGES_REQUESTED** [review]
@jonbodner Can you update the `.github/CODEOWNERS` and add your team as the owner of this check (see this example [here](https://github.com/DataDog/datadog-agent/blob/main/.github/CODEOWNERS#L322))

**Comment by schneideradam** [comment]
@rarguelloF 
From @carlosroman 
```
It would be good to get a benchmark showing it enabled and disabled to better understand the overhead.)
Rather than find out during RC deployments
```

**Comment by jonbodner** [comment]
> @jonbodner Can you update the `.github/CODEOWNERS` and add your team as the owner of this check (see this example [here](https://github.com/DataDog/datadog-agent/blob/main/.github/CODEOWNERS#L322))

@carlosroman do I include that in this PR, or in a new one?

**Review by hush-hush — APPROVED** [review]
Looks good for files own by ASC

**Review by jonbodner — APPROVED** [review]

File: pkg/collector/corechecks/servicediscovery/events.go
Comment (yshapiro-57) [inline pkg/collector/corechecks/servicediscovery/events.go:31]: @rarguelloF I talked to @jonbodner yesterday and he confirmed that he is OK with removing api_version and request_type from the payload and only keeping them in the main event struct.

**Review by jszwedko — COMMENTED** [review]
Approving for files owned by @DataDog/agent-processing-and-routing

**Review by jszwedko — APPROVED** [review]

**Review by carlosroman — APPROVED** [review]

File: pkg/collector/corechecks/servicediscovery/events.go
Comment (rarguelloF) [reply pkg/collector/corechecks/servicediscovery/events.go]: removed those fields 👍

**Review by yshapiro-57 — APPROVED** [review]
I am not an expert on the process agent, but the code that sends events to the instrumentation telemetry backend looks legit to me. So I am approving.

**Review by spencergilbert — APPROVED** [review]
Approved the one BaRX owned file

**Pull request merged**
